File: MaxSubarray.py
# -*- coding: utf-8 -*-
import sys
import time


class MaxSubarray:

    def __init__(self, vetor):
        string_resultado = "VETOR TAMANHO "+str(len(vetor))
        
        string_resultado += "\nDynamic Programing -> "
        inicio = time.clock()
        res = self.dynamicPrograming(vetor)
        total = time.clock() - inicio
        string_resultado += str(int(total*1000)) + " ms"

        string_resultado += "\nDivide and Conquerer -> "
        inicio = time.clock()
        res = self.divideAndConquer(vetor, 0, len(vetor)-1)
        total = time.clock() - inicio
        string_resultado += str(int(total*1000)) + " ms"

        string_resultado += "\nBetter Enumeration -> "
        inicio = time.clock()
        res = self.betterEnumeration(vetor)
        total = time.clock() - inicio
        string_resultado += str(int(total*1000)) + " ms"

        string_resultado += "\nEnnumeration -> "
        inicio = time.clock()
        res = self.enumeration(vetor)
        total = time.clock() - inicio
        string_resultado += str(int(total*1000)) + " ms"
        string_resultado += "\n"

        print string_resultado


    def  dynamicPrograming(self, vetor) :
        maior = vetor[0]
        tam = len(vetor)
        soma = [tam]
        soma[0] = vetor[0]

        for i in range(1, tam):
            soma.append(max(vetor[i], (soma[i - 1] + vetor[i])))
            maior = max(maior, soma[i])
        
        return maior


    def  somaSubvetor(self, vetor,  primeiro,  meio,  ultimo) :

        somaEsquerda = 0
        soma = 0

        for i in range(meio, primeiro, -1): 
            soma += vetor[i]
            somaEsquerda = max(somaEsquerda, soma)
        

        somaDireita = 0
        soma = 0
        for i in range(meio+1, ultimo):
            soma += vetor[i]
            somaDireita = max(somaDireita, soma)
        

        return somaEsquerda + somaDireita
    

    def  maiorDe3(self, a,  b,  c) :
        if (a > b and a > c) :
            return a
        elif (b > a and b > c) :
            return b
        else :
            return c
        
    
    def  divideAndConquer(self, vetor,  primeiro,  ultimo): 
        if (primeiro == ultimo) :
            return vetor[primeiro]
        else : 
            meio = (primeiro + ultimo) / 2 
          
            a = self.divideAndConquer(vetor, primeiro, meio)
            b = self.divideAndConquer(vetor, meio + 1, ultimo)
            c = self.somaSubvetor(vetor, primeiro, meio, ultimo)

            return self.maiorDe3(a, b, c)
    

    def  betterEnumeration(self, vetor) :
        tam = len(vetor)
        somaAtual = 0
        maiorSoma = 0

        for i in range(0, tam):
            for j in range(i+1, tam):
                somaAtual = vetor[i]
                for k in range(i+1, j): 
                    somaAtual += vetor[k]
                
                maiorSoma = max(somaAtual, maiorSoma)
            
        
        return maiorSoma


    def  enumeration(self, vetor):
        tam = len(vetor)
        maiorSoma = vetor[0]

        for i in range(0, tam):
            for j in range(i, tam):
                somaAtual = 0
                for k in range(i, j):
                    somaAtual += vetor[k]
                
                maiorSoma = max(somaAtual, maiorSoma)
        
        return maiorSoma
    

    
    

if __name__ == '__main__':
    #vetor = [31, -41, 59, 26, -53, 58, 97, -93, -23, 84]
    print "#################"
    print "PYTHON"

    caminho = "/home/jennifer/Documents/UTFPR/6º Período/Análise -  3ª triste tentativa/APS/APSAnalise/src/Main/vetores.txt"
    arq = open(caminho, 'r')
    texto = arq.read()
    linha = ""
    for c in texto:
        if c == '\n':
            vetor = []
            vs = linha.split(" ")
            linha = ""

            for x in vs:
                try:
                    vetor.append(int(x))
                except:
                    print ""

            MaxSubarray(vetor)
        else:
            linha += c

    arq.close()
    

File: MaxSubarray.rb

class MaxSubarrayAlgorithms

    def max(a, b)
        if a > b 
            return a
        else
            return b
        end
    end

    def dynamicPrograming(vetor)
        maior = vetor[0]
        soma = [vetor.length]
        soma[0] = vetor[0]
        i = 1
        while i < vetor.length
            soma[i] = max(vetor[i], soma[i - 1] + vetor[i])
            maior = max(maior, soma[i])
            i += 1
        end
        return maior
    end

    def somaSubvetor(vetor, primeiro, meio, ultimo)
        somaEsquerda = 0
        soma = 0
        i = meio
        while i >= primeiro
            soma += vetor[i]
            somaEsquerda = max(somaEsquerda, soma)
            i -= 1
        end
        somaDireita = 0
        soma = 0
        i = meio + 1
        while i <= ultimo
            soma += vetor[i]
            somaDireita = max(somaDireita, soma)
            i += 1
        end
        return somaEsquerda + somaDireita
    end

    def maiorDe3(a, b, c)
        if a > b and a > c then
            return a
        elsif b > a and b > c then
            return b
        else
            return c
        end
    end

    def divideAndConquer(vetor, primeiro, ultimo)
        if primeiro == ultimo then
            return vetor[primeiro]
        else
            meio = (primeiro + ultimo) / 2
            return self.maiorDe3(self.divideAndConquer(vetor, primeiro, meio), self.divideAndConquer(vetor, meio + 1, ultimo), self.somaSubvetor(vetor, primeiro, meio, ultimo))
        end
    end

    def betterEnumeration(vetor) 
        tam = vetor.length
        somaAtual = 0
        maiorSoma = 0

        i = 0
        while i < tam
            j = i+1
            while j < tam
                somaAtual = vetor[i]
                k = i+1
                while k < j
                    somaAtual += vetor[k]
                    k += 1
                end
                maiorSoma = max(somaAtual, maiorSoma)
                j += 1
            end
            i += 1
        end  
        
        return maiorSoma
    end

    def enumeration(vetor)
        tam = vetor.length
        maiorSoma = vetor[0]

        i = 0
        while i < tam
            j = i
            while j < tam
                somaAtual = 0
                k = i
                while k < j
                    somaAtual += vetor[k]
                    k +=1
                end
                maiorSoma = max(somaAtual, maiorSoma)
                j += 1
            end
            i += 1
        end
        return maiorSoma
    end

    def executar(vetor)
        string_resultado = "\nVETOR TAMANHO "+vetor.length.to_s

        string_resultado += "\nDynamic Programing -> "
        inicio = Time.now
        res = self.dynamicPrograming(vetor)
        total = Time.now - inicio
        string_resultado += (total*1000).to_i.to_s + " ms"

        string_resultado += "\nDivide and Conquerer -> "
        inicio = Time.now
        res = self.divideAndConquer(vetor, 0, vetor.length-1)
        total = Time.now - inicio
        string_resultado += (total*1000).to_i.to_s + " ms"

        string_resultado += "\nBetter Enumeration -> "
        inicio = Time.now
        res = self.betterEnumeration(vetor)
        total = Time.now - inicio
        string_resultado += (total*1000).to_i.to_s + " ms"

        string_resultado += "\nEnnumeration -> "
        inicio = Time.now
        res = self.enumeration(vetor)
        total = Time.now - inicio
        string_resultado += (total*1000).to_i.to_s + " ms"
        string_resultado += "\n"


        print string_resultado

    end

end

vetor = [31, -41, 59, 26, -53, 58, 97, -93, -23, 84]
msa = MaxSubarrayAlgorithms.new

caminho = "/home/jennifer/Documents/UTFPR/6º Período/Análise -  3ª triste tentativa/APS/APSAnalise/src/Main/vetores.txt"

puts "#################"
puts "RUBY"

File.open(caminho, 'r') do |f1|
    while linha = f1.gets
        vetor = []
        vs = linha.split(" ")
        linha = ""
        i = 0
        for x in vs
                vetor[i] = x.to_i
                i += 1
        end
        msa.executar(vetor)
    end
end
